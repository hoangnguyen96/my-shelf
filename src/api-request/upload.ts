import { MESSAGES } from "@app/constants";

export const generateImageUpload = async (file: File) => {
  try {
    const formData = new FormData();
    formData.append("image", file);

    const response = await fetch(
      `https://api.imgbb.com/1/upload?key=${process.env.NEXT_PUBLIC_IMGBB_API_KEY}`,
      {
        method: "POST",
        body: formData,
      }
    );

    if (!response.ok) {
      throw new Error(MESSAGES.UPLOAD_FAILED);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw new Error(MESSAGES.ERROR_UPLOAD);
  }
};
