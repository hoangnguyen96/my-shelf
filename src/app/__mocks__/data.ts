export const DATA_BOOKS = [
  {
    title: "Don't Make Me Think",
    author: "Steve Krug",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/QbQgVtG/book1.png",
    description: "description will update later",
    status: false,
    publicationYear: 2000,
    rating: 4.5,
    createdDate: "3 Sep 2024 9:07 PM",
    edition: "first",
    id: "1",
  },
  {
    title: "Rich Dad Poor Dad",
    author: "Robert T. Kiyosaki",
    category: "SelfHelpBook",
    imageUrl: "https://i.ibb.co/9qHTtQF/book2.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1997,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:47 PM",
    edition: "second",
    id: "2",
  },
  {
    title: "The Design of Everyday Things",
    author: "Don Norman",
    category: "TextBook",
    imageUrl: "https://i.ibb.co/LnSBcCM/book5.png",
    description:
      "Steve Krug is a usability consultant who has more than 30 years of experience as a user advocate for companies like Apple, Netscape, AOL, Lexus, and others. Based in part on the success of his first book, Don't Make Me Think, he has become a highly sought-after speaker on usability design.",
    status: false,
    publicationYear: 1988,
    rating: 4.5,
    createdDate: "3 Sep 2024 8:51 PM",
    edition: "fourth",
    id: "3",
  },
];

export const DATA_USER = [
  {
    username: "admin",
    email: "admin@gmail.com",
    password: "$2a$10$MwSZ9vWSX7DVJwJ4SF5Mzuj2zMFjY52XM9L11gGG0j79gybowaLfq",
    isAdmin: true,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/RHMqQGr/man-1.png",
    favorites: ["3", "2", "1", "13", "5"],
    shelfBooks: ["1", "6"],
    userId: "3733403",
    id: "1",
  },
  {
    username: "hung",
    email: "hung1@gmail.com",
    password: "$2a$10$Mt6ruwOzt3nzX1.HpSt.OOzut.teXmLE74WclmkP.Nn.gzVQrlqri",
    isAdmin: false,
    phone: "",
    bio: "",
    avatar: "https://i.ibb.co/88X1WfZ/avatar-default.png",
    favorites: ["1"],
    shelfBooks: ["2", "3", "1"],
    userId: "8857134",
    id: "2",
  },
];
