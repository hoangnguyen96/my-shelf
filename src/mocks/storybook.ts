export const mockRouter = {
  pathname: "/",
  route: "/",
  query: {},
  asPath: "/",
  basePath: "",
  push: () => Promise.resolve(true),
  replace: () => Promise.resolve(true),
  reload: () => {},
  back: () => {},
  prefetch: () => Promise.resolve(),
  beforePopState: () => {},
  isFallback: false,
  isLocaleDomain: false,
  isReady: true,
  isPreview: false,
  forward: () => {},
  events: {
    on: () => {},
    off: () => {},
    emit: () => {},
  },
};

export const mockSession = {
  user: {
    isAdmin: true,
    email: "admin@gmail.com",
    id: "3733403",
    name: "admin",
    image: "https://i.ibb.co/RHMqQGr/man-1.png",
  },
  expires: "2024-12-31T23:59:59.999Z",
};
