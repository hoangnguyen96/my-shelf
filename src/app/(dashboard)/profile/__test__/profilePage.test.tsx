import { act, fireEvent, render } from "@testing-library/react";
import { useSession } from "next-auth/react";
import { getUserById } from "@app/api-request";
import { DATA_USER } from "@app/mocks/data";
import ProfilePage from "../page";

jest.mock("next-auth/react", () => ({
  useSession: jest.fn(),
}));

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@app/api-request", () => ({
  getUserById: jest.fn(),
}));

jest.mock("@app/actions/auth", () => ({
  logout: jest.fn(),
}));

describe("Profile Page", () => {
  (useSession as jest.Mock).mockReturnValue({
    data: {
      user: {
        isAdmin: true,
        email: "admin@gmail.com",
        id: "3733403",
        name: "admin",
        image: "https://i.ibb.co/RHMqQGr/man-1.png",
      },
      expires: "2024-12-31T23:59:59.999Z",
    },
    status: "authenticated",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    (getUserById as jest.Mock).mockReturnValue({
      dataUserById: DATA_USER[0],
    });
  });

  it("Should render correctly snapshot", async () => {
    (useSession as jest.Mock).mockReturnValue({
      data: {
        user: {
          isAdmin: true,
          email: "admin@gmail.com",
          id: "3733403",
          name: "admin",
          image: "",
        },
        expires: "2024-12-31T23:59:59.999Z",
      },
      status: "authenticated",
    });

    await act(async () => {
      const { container } = render(<ProfilePage />);
      expect(container).toMatchSnapshot();
    });
  });

  it("Should handle getUserById error", async () => {
    (getUserById as jest.Mock).mockRejectedValue(
      new Error("Failed to fetch books")
    );
    await act(async () => {
      const { container } = render(<ProfilePage />);
      expect(container).toMatchSnapshot();
    });
  });

  // TODO: Will update later
  // it("should handle update user", async () => {
  //   (getUserById as jest.Mock).mockResolvedValue({
  //     dataUserById: DATA_USER[0],
  //   });
  //   const { container, getByRole, getByPlaceholderText, getByTestId } = render(
  //     <ProfilePage />
  //   );

  //   await act(async () => {

  //     // Click the 'Edit' button to make the form editable
  //     const editButton = getByTestId("click-un-read-only");
  //     fireEvent.click(editButton);

  //     // Fill out the form with new data
  //     fireEvent.change(getByPlaceholderText("Full name"), {
  //       target: { value: "Jane Doe" },
  //     });
  //     fireEvent.change(getByPlaceholderText("Email..."), {
  //       target: { value: "jane.doe@example.com" },
  //     });
  //     fireEvent.change(getByPlaceholderText("Phone"), {
  //       target: { value: "0987654321" },
  //     });
  //     fireEvent.change(getByPlaceholderText("Reason For Your Contribution"), {
  //       target: { value: "Junior Developer" },
  //     });

  //     // Click the submit button to submit the form
  //     const submitButton = getByRole("button", { name: /update profile/i });
  //     fireEvent.click(submitButton);

  //     expect(submitButton).toBeDisabled();
  //   });
  // });
});
