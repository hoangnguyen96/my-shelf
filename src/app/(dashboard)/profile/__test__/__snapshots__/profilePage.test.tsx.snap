// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Page Should handle getUserById error 1`] = `<div />`;

exports[`Profile Page Should render correctly snapshot 1`] = `<div />`;
